internal/k8s/controllers/layer2_status_controller.go

```diff
@@ -0,0 +1,127 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"go.universe.tf/metallb/api/v1beta1"
+	"go.universe.tf/metallb/internal/layer2"
+)
+
+type StatusFetcher func(types.NamespacedName) []layer2.IPAdvertisement
+
+type l2StatusEvent struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+}
+
+func (evt *l2StatusEvent) DeepCopyObject() runtime.Object {
+	res := new(l2StatusEvent)
+	res.Name = evt.Name
+	res.Namespace = evt.Namespace
+	return res
+}
+
+func NewL2StatusEvent(namespace, name string) event.GenericEvent {
+	evt := l2StatusEvent{}
+	evt.Name = name
+	evt.Namespace = namespace
+	return event.GenericEvent{Object: &evt}
+}
+
+type Layer2StatusReconciler struct {
+	client.Client
+	Logger        log.Logger
+	NodeName      string
+	ReconcileChan <-chan event.GenericEvent
+	// fetch ipAdv object to get interface info
+	StatusFetcher StatusFetcher
+}
+
+func (r *Layer2StatusReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "Layer2StatusReconciler", "start reconcile", req.NamespacedName.String())
+	defer level.Info(r.Logger).Log("controller", "Layer2StatusReconciler", "end reconcile", req.NamespacedName.String())
+
+	ipAdvS := r.StatusFetcher(req.NamespacedName)
+
+	objName := fmt.Sprintf("%s-%s", req.Name, r.NodeName)
+	if len(ipAdvS) == 0 {
+		err := r.Client.Delete(ctx, &v1beta1.ServiceL2Status{ObjectMeta: metav1.ObjectMeta{Name: objName, Namespace: req.Namespace}})
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	state := &v1beta1.ServiceL2Status{ObjectMeta: metav1.ObjectMeta{
+		Name:      objName,
+		Namespace: req.Namespace,
+	}}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: objName, Namespace: req.Namespace}, state)
+	if err != nil && !errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	desiredState := r.buildDesiredStatus(objName, req.Namespace, req.Name, ipAdvS)
+	if reflect.DeepEqual(state, desiredState) {
+		return ctrl.Result{}, nil
+	}
+
+	var result controllerutil.OperationResult
+	if result, err = controllerutil.CreateOrPatch(ctx, r.Client, state, func() error {
+		state.Labels = desiredState.Labels
+		state.Status = desiredState.Status
+		return nil
+	}); err != nil {
+		return ctrl.Result{}, err
+	} else if result == controllerutil.OperationResultCreated {
+		// According to https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrPatch
+		// If the object is created for the first time, we have to requeue it to ensure that the status is updated.
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	level.Debug(r.Logger).Log("controller", "Layer2StatusReconciler", "updated state", dumpResource(state))
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Layer2StatusReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1beta1.ServiceL2Status{}).
+		WatchesRawSource(&source.Channel{Source: r.ReconcileChan}, &handler.EnqueueRequestForObject{}).
+		Complete(r)
```

Comment:
we can now add a `WithEventFilter` to reconcile only those that end with the node's name